### ISSUE TITLE
Reduce memory usage for immutable documents

### ISSUE THREAD

**Issue by annulen** [zeux/pugixml#50]
Work in this branch was not finished: https://code.google.com/p/pugixml/source/browse/#svn%2Fbranches%2Fimmutable
Is there any chance for it to be completed?

**Comment by zeux** [comment]
The original idea that I had for immutable tree storage proved too cumbersome to implement.

The idea was to byte-pack everything using offsets and dynamically figuring out the best encoding based on the required offset length etc. - e.g. if the parent node data is close to the child node data we can pack parent pointer in 1 byte; if it's further we may use 2 bytes etc.

While initial prototypes showed promising results in terms of resulting size, there were several complications.

The first issue is that it's actually impossible to determine the size of forward references beforehand - e.g. "next sibling" could be arbitrarily far down the byte stream depending on the amount of data in the subtree. This means that we either have to plan for the worst case, reducing efficiency, allocate worst-case and compact at some points during parsing which is possible but further complicates the parsing, or optimistically use small storage and have some way to fall back to a slower data structure, so access is not guaranteed O(1) any more (but may be amortized O(1) depending on the chosen structure).

The second complication, which was the primary reason for abandoning this idea, was that it became very hard to keep most of the codebase the same. It would be possible to significantly rewrite the code to use the byte stream approach, but that results in a forked codebase; trying to merge the two approaches became too complicated because the concept of "node object" is no longer there; additionally to support some read-only functions the size of xml_node/xml_attribute objects would have to grow (which is not a memory issue but yet another ABI change).

Finally, the design of the byte stream itself is full of tradeoffs that are hard to make without knowing anything about the document structure. I.e. originally I thought that "first_child" would just use 1 bit in the node header, since the node is right after the current one in the stream. Unfortunately, this only works if you either split attributes into a separate stream or skip attributes when looking for first child (so first_child() is not O(1) anymore); additionally, given a varying size of the node these operations require parsing the node header fully which is not terribly fast or requires a specially crafted storage structure and a look up table...

**Comment by zeux** [comment]
This was the bad news.

The good news is that a different idea is already implemented in trunk and will ship with 1.7.

pugixml now supports "compact mode" - it's enabled by defining PUGIXML_COMPACT. It does not change ABI or functionality in any way - everything that you could do before you can still do. All operations are amortized O(1) - usually for most nodes all of them are O(1) (although of course the constant is slightly higher than for non-compact mode because you can't beat pointer access as long as data is in cache) but there could be some outliers that take slightly longer to query data about.

The node size in this mode is 12 bytes (instead of 40/80 bytes on x86/x64); the attribute size is 8 bytes (instead of 20/40 bytes on x86/x64). In addition to that there's a separate data structure that is sometimes used to store references that don't fit in the base structure; for most documents this data structure is small.

You're free to modify the tree. Some modifications still result in the same (low) memory consumption, e.g. appending a lot of nodes should still be fast. Some modifications may force the storage to revert to using the data structure more and more. It depends.

Parsing is usually slightly slower in this mode, as well as the tree traversal because the individual operations are slightly slower - the slowdown is not very big usually, and for some documents you actually see a performance improvement because there's way less memory to access.

**Comment by zeux** [comment]
Finally, the new structure should be compatible with the patch that removes PCDATA from the document on parse by storing the value pointer inside the node object. I have not tried this but it should work and should not deteriorate the efficiency.

All of the above will be documented at some point before 1.7 gets released :) If you get a chance to use this please let me know whether it works well for your use case.

In a nutshell, compact mode is much easier for me to maintain, does not require immutability - but it can be less efficient than a byte stream solution because the nodes are forced to be the same size and there are no constraints for how the nodes refer to each other.

**Comment by annulen** [comment]
Thank you! I've upgraded my copy of pugixml to current master (see #51; it was too hard for me to solve some conflicts so I used bare minimum from pcdata patch). According to callgrind instruction count on my desktop machine, new version should be ~10% faster than pugixml 1.2 in my workload. I've also tried PUGIXML_COMPACT mode - it doesn't break my code, but it's about 2x slower than regular mode - this means to keep speed high I'd have to use 2 separate copies of pugixml in my application, regular for small files and compact for large ones.

Does compact mode rely on unaligned access somewhere? Most of my targets don't provide efficient unaligned access.

**Comment by zeux** [comment]
Compact mode does not rely on unaligned access.

I would be surprised to see 2x slowdown. I'm not sure instruction count is an accurate representation of performance, and I've only measured performance on x86 and x64 so it's possible I guess that on some ARM processors the gap between regular mode and compact is wider.

Here's some performance data from files that are reasonably heavy on the amount of markup:

x64, non-compact:
data/employees-big.xml: 14.536 msec, 39.750 Mb (9.999 + 29.750), 432361 nodes, 84070 attributes
data/terrover.xml: 15.675 msec, 34.581 Mb (15.771 + 18.810), 162389 nodes, 230780 attributes
data/vod.xml: 3.382 msec, 10.578 Mb (2.082 + 8.496), 138305 nodes, 6 attributes

x86, non-compact:
data/employees-big.xml: 16.863 msec, 24.850 Mb (9.999 + 14.850), 432361 nodes, 84070 attributes
data/terrover.xml: 19.715 msec, 25.170 Mb (15.771 + 9.399), 162389 nodes, 230780 attributes
data/vod.xml: 3.110 msec, 6.343 Mb (2.082 + 4.261), 138305 nodes, 6 attributes

x64, compact:
data/employees-big.xml: 17.697 msec, 15.652 Mb (9.999 + 5.653), 432361 nodes, 84070 attributes
data/terrover.xml: 18.466 msec, 19.686 Mb (15.771 + 3.914), 162389 nodes, 230780 attributes
data/vod.xml: 4.378 msec, 3.805 Mb (2.082 + 1.722), 138305 nodes, 6 attributes

**Comment by zeux** [comment]
callgrind reports that the instruction count increases almost 2x on "vod.xml" on x64 (24M -> 44M).

**Comment by zeux** [comment]
Although I guess everything above is just parsing so it could be that if access time is dominant the actual slowdown is more significant.

**Comment by annulen** [comment]
>I'm not sure instruction count is an accurate representation of performance
/me too, but from my experience it ususally shows the trends right, and is very reproducible (one binary in one testing scenario produces the same instruction count almost exactly)

**Comment by kirbyfan64** [comment]
@annulen Instruction count is a *horrible* speed metric. This:

```
strcpy:
  rep movsb
  ret
```

is slower than [this huge, 1000-line thing](https://github.com/android/platform_bionic/blob/master/libc/arch-x86_64/string/sse2-strcpy-slm.S).

**Comment by annulen** [comment]
@kirbyfan64 Callgrind counts instructions with rep prefix as separate repeated instructions, and I have no SIMD involved

**Comment by zeux** [comment]
Right, easy reproducibility is a big bonus; I'm just saying that it's possible that the ratio of instruction count is very different from the ratio of actual performance. In some cases even the trend can be different (although I guess these are less common).

I'll take a look at code that gcc produces to figure out if I can optimize it; I spent some time tuning compact mode to produce faster code with clang, but it looks like gcc output is very different.

**Comment by zeux** [comment]
The code generated by gcc looks reasonably close to clang. gcc does not do integral promotion to 64-bit integers for parts of compact decoding which seems like the largest source of discrepancy but it's not clear whether it actually has a performance cost.

Since I don't really see ways to noticeably improve compact mode performance at this point, I'll assume it's reasonably close to being optimal :) it may get slightly better over time if I get interesting ideas, similar to how core parser got a bit better over time.

The summary from the above comments made it into documentation in trunk: https://rawgit.com/zeux/pugixml/master/docs/manual.html#dom.memory.compact, so I'm closing this issue.
